### ISSUE TITLE
Bicubic (and biquadratic) interpolation at the edges of grids

### ISSUE THREAD

**Issue by ccrook** [opengeospatial/CRS-Gridded-Geodetic-data-eXchange-Format#21]
The original text of this issue is below - however the discussion entirely relates to definition of uncertainty so I have renamed this issue to match.

---

I plan to add bicubic and biquadratic interpolation to the GGXF.py module I am constructing at https://github.com/opengeospatial/CRS-Gridded-Geodetic-data-eXchange-Format/tree/script_development/scripts

Which leads me to wonder what our specification is for evaluating these functions at the edges of grids.  eg bicubic used 16 nodes from 9 grid cells centred on the evaluation point. If the evaluation point is in a grid cell on the edge of the grid then three of these cells are not on the grid at all.  A similar situation occurs for biquadratic interpolation (which incidentally I would definitly not encourage using)

How should the model be evaluated at these points?  One simple option is to use the nearest 9 cells.  Do we have a preferred approach (and/or a reference to one).

**Comment by ccrook** [comment]
To be defined by citation in GGXF specification (2 Mar 2022 discussion)

**Comment by kevinmkelly** [comment]
As requested, I have provided citations and draft definitions for the interpolation methods in Table C.10. I have also proposed five versions of a definition for "uncertaintyMeasure" in Table C.2. These are given in this Word document:
[citations+uncertainty.docx](https://github.com/opengeospatial/CRS-Gridded-Geodetic-data-eXchange-Format/files/8227522/citations%2Buncertainty.docx)
.

**Comment by ccrook** [comment]
@kevinmkelly For ease of comment I have copied the proposed definitions here:

>  Definition:  
    1. Non-negative measure of the expected variation of the parameter.  
    2. Numerical value quantifying the expected variation of the parameter.  
    3. Numerical range within which it is expected the parameter will fall.  
    4. Expresses the statistical dispersion of the parameter.  
    5. A numerical value which gives the best estimate of the parameter.  
The uncertainty measure may be separated into horizontal and vertical components.

My first thought is that 3 would defined by two values rather than one, and that 5 describes the parameter itself rather than the uncertainty.  I wonder if this definition needs to specify that there is a physical quantity that the gridded data is representing, so taking some of the words from the other definitions something like:

> a statistical measure of the unknown difference between the value of a parameter at any location and the true value of the physical quantity it represents

I am wary of talking about the variation of the parameter, as the GGXF is a single realisation.  The parameter doesn't vary.  It is constant, but probably a bit wrong.

I am not sure that "statistical measure" is a good terminology either.  Quite likely it is not a statistical measure but just a number that feels about right, so maybe something more like "an indicative measure" would be better?

**Comment by kevinmkelly** [comment]
@ccrook: Yes, correct, definition 5 is not suitable. Definition 3 does not need to have two values but the definition would need to be expanded, i.e. a standard deviation is a single value that defines a range. So, we could revise 3 to read something like:

> 3. A numerical value that when added to and subtracted from the parameter defines an interval within which the parameter is expected to fall.

Regardless of whether the GGXF is a single realization or not, the values, or parameters as we are calling them in the specification, are statistical quantities estimated from measurements. Thus, all the values/parameters in a GGXF grid are "random variables" - even an uncertainty grid is a grid of random variables! So, your definition could read (I don't think "unknown difference" correct here):

> A statistical measure defining an interval within which the true value of a parameter is expected to fall.

or

> A numerical range defined as an interval with a lower bound and an upper bound relative to the parameter within which the true value of a parameter is expected to fall.

The first may be too terse and not precise enough and the second may be too wordy and clunky, but slightly more explicit.

I think we should assume users have enough knowledge of basic statistics that we need not be too verbose - that sometimes obfuscates what we are truing to say. On the other hand, too few words can make the definition unclear and/or ambiguous. Hopefully, between us and others we can capture the right balance!

**Comment by ccrook** [comment]
Hi Kevin.   We may be encroaching on philosophy rather than mathematics here.  I guess what I am saying is, yes this grid is a single sample from an infinite set of possible grids we could have ended up with.  The process of deriving them includes a number of factors which introduce randomness, as well as some (such as using interpolation on a grid) that introduce systematic errors.  However the parameters themselves are no longer random once they have selected it from, it is the probability space they are selected from that is random.  Once the dice have landed they don't change.

On the assumption of knowledge - this is a perennial problem with definition in my mind.  It is pretty safe to assume that anyone reading this will know what all these things are if it matters to them, and won't care if it doesn't.  (For example a coder doesn't need to know what an uncertainty is, just that it is a floating point number).   In reality all we need to clarify is where there might be ambiguity in usage (eg between grid north and true north for a displacement - possibly better not to use north at all, or whether we are rotating coordinates or axes if that were something we were doing).

So I think the things that we need to encapsulate are that this is a representation of our expectation of the error of the parameter.  This requires the notion of a "difference from a true value".  And it makes no sense to use our value if we know the true value, so "unknown true value".  

I suspect we will really have a very realistic measure of this uncertainty - as you say it has its own uncertainty.  So calling it "statistical" may be overselling it.  Which gets me to:

> an indicative measure of the magnitude of the difference between the value of a parameter at any location and the unknown true value of the physical quantity it represents

Yes it is a bit verbose, but it is still only one sentence.  However most words are there for a reason and carry some of the overall concept of the uncertainty.  The word "measure" is the perhaps the worst.  It isn't measured and doesn't measure anything.  Maybe it would be better to bring "unknown" forward, so to reduce verbosity a bit

> represents the unknown magnitude of the difference between the value of a parameter and the true value of the physical quantity it is an estimation of

... enough rambling!

**Comment by kevinmkelly** [comment]
@ccrook: No worries. Though, I think there will be very, very knowledgeable folks reading and interpreting our specification who will know quite a bit about the subject matter (and statistics). They may the ones that will guide developers. Anyway, no need to quibble any further; I'll propose:

> A numerical measure of the magnitude of the difference between the value of a parameter and the unknown true value of the physical quantity it represents.

I don't see the need for "at any location"; the parameters come from the grid at fixed locations at every grid node. But I won't fuss if you feel it needs to be there.

**Comment by ccrook** [comment]
This issue has concluded the discussion on description of uncertainty...
